pebbles/views/namespaced_keyvalues.py

```diff
@@ -0,0 +1,123 @@
+from pebbles.server import restful
+from pebbles.views.commons import auth
+from pebbles.utils import requires_admin
+
+namespaced_keyvalues = Blueprint('namespaced_keyvalues', __name__)
+
+
+namespace_fields = {
+    'id': fields.String(attribute='id'),
+    'namespace': fields.String,
+    'key': fields.String,
+    'value': fields.Raw,
+    'created': fields.Float,
+    'updated': fields.Float
+}
+
+
+class NamespacedKeyValueList(restful.Resource):
+    parser = reqparse.RequestParser()
+    parser.add_argument('namespace', type=str)
+    parser.add_argument('key', type=str)
+
+    @auth.login_required
+    @requires_admin
+    @marshal_with(namespace_fields)
+    def get(self):
+        args = self.parser.parse_args()
+        namespaced_query = NamespacedKeyValue.query
+        if args.get('namespace'):
+            namespaced_query = namespaced_query.filter_by(namespace=args.namespace)
+        if args.get('key'):
+            namespaced_query = namespaced_query.filter_by(key=args.key)
+
+        return namespaced_query.all()
+
+    @auth.login_required
+    @requires_admin
+    def post(self):
+        form = NamespacedKeyValueForm()
+
+        if not form.validate_on_submit():
+            logging.warn(form.errors)
+            logging.warn("validation error on creating namespaced key value data")
+            return form.errors, 422
+
+        namespace = form.namespace.data
+        key = form.key.data
+        value = form.value.data
+
+        ns_check = NamespacedKeyValue.query.filter_by(namespace=namespace, key=key).first()
+        if ns_check:
+            logging.warn("a combination of namespace %s with key %s already exists" % (namespace, key))
+            abort(422)
+        namespaced_keyvalue = NamespacedKeyValue(namespace, key)
+        namespaced_keyvalue.value = value
+
+        curr_ts = time.time()
+        namespaced_keyvalue.created_ts = curr_ts
+        namespaced_keyvalue.updated_ts = curr_ts
+
+        db.session.add(namespaced_keyvalue)
+        db.session.commit()
+
+
+class NamespacedKeyValueView(restful.Resource):
+    parser = reqparse.RequestParser()
+
+    @auth.login_required
+    @requires_admin
+    @marshal_with(namespace_fields)
+    def get(self, namespace, key):
+        namespaced_keyvalue = NamespacedKeyValue.query.filter_by(namespace=namespace, key=key).first()
+        if not namespaced_keyvalue:
+            abort(404)
+
+        return namespaced_keyvalue
+
+    @auth.login_required
+    @requires_admin
+    def put(self, namespace, key):
+        form = NamespacedKeyValueForm()
+
+        if not form.validate_on_submit():
+            logging.warn(form.errors)
+            logging.warn("validation error on modifying namespaced key value data")
+            return form.errors, 422
+
+        value = form.value.data
+        updated_version_ts = float(form.updated_version_ts.data)
+
+        namespaced_keyvalue_query = NamespacedKeyValue.query.filter_by(namespace=namespace, key=key)
```

Comment:
adding with_for_update to the query should fix any really tight race conditions,

see https://www.postgresql.org/docs/9.0/static/sql-select.html#SQL-FOR-UPDATE-SHARE
and http://docs.sqlalchemy.org/en/latest/orm/query.html#sqlalchemy.orm.query.Query.with_for_update